### PR TITLE
ENT-4653 ifelse() not evaluated in case of unresolved vars

### PR DESCRIPTION
**Pull request by craigcomstock**

### CODE REVIEW

**Comment by craigcomstock** [comment]
@olehermanse @vpodzime what do you think of this PR? It is passing tests on travis. Just want to get a gut level reaction to the strategy. This PR will almost certainly supercede https://github.com/cfengine/core/pull/4141

**Comment by craigcomstock** [comment]
@cf-bottom jenkins with exotics please.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5031/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5031/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5031/

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5035/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5035/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5035/

File: libpromises/expand.c [hunk 1]
```diff
@@ -188,13 +188,15 @@ static void MapIteratorsFromRval(EvalContext *ctx,
...
+                                        PromiseActuator *act_on_promise, void *param,
+                                        bool do_ifelse)
 {
     PromiseResult result = PROMISE_RESULT_SKIPPED;
 
     /* TODO this loop could be completely skipped for for non vars/classes if
      *      act_on_promise is CommonEvalPromise(). */
-    while (PromiseIteratorNext(iterctx, ctx))
+    bool done = ! do_ifelse;
+    while (PromiseIteratorNext(iterctx, ctx) || !done)
```
Comment (vpodzime) [inline libpromises/expand.c:199]: `|| !done`? Shouldn't the loop stop when done no matter what?

File: libpromises/expand.c [hunk 2]
```diff
@@ -247,6 +251,7 @@ static PromiseResult ExpandPromiseAndDo(EvalContext *ctx, PromiseIterator *iterc
 PromiseResult ExpandPromise(EvalContext *ctx, const Promise *pp,
                             PromiseActuator *act_on_promise, void *param)
 {
+    bool do_ifelse = false;
```
Comment (vpodzime) [inline libpromises/expand.c:254]: Please define the variable near to the place it it's used for the first time.

File: libpromises/expand.c [hunk 3]
```diff
@@ -188,13 +188,15 @@ static void MapIteratorsFromRval(EvalContext *ctx,
...
-                                        PromiseActuator *act_on_promise, void *param)
+                                        PromiseActuator *act_on_promise, void *param,
+                                        bool do_ifelse)
 {
     PromiseResult result = PROMISE_RESULT_SKIPPED;
 
     /* TODO this loop could be completely skipped for for non vars/classes if
      *      act_on_promise is CommonEvalPromise(). */
-    while (PromiseIteratorNext(iterctx, ctx))
+    bool done = ! do_ifelse;
```
Comment (vpodzime) [inline libpromises/expand.c:198]: This translation from `do_ifelse` to `done` is very confusing and deserves a big comment. Or maybe actually working with `do_ifelse` would make things clearer?

**Comment by craigcomstock** [comment]
> It's very hard to see how this resolves the issue (even though I believe it does).

I agree. I will try to re-work it. The basic idea is this: if the promise involves an ifelse expression it should be evaluated/actuated even if there are unresolved parameters. In this case I decided to force evaluation even when PromiseIteratorNext() returns false which would normally prevent evaluation from happening.

**Comment by craigcomstock** [comment]
@vpodzime I believe I have addressed your concerns. I am going to test the acceptance test locally without the C change right now to make sure it "catches" the problem.

File: libpromises/expand.c [hunk 4]
```diff
@@ -188,13 +188,21 @@ static void MapIteratorsFromRval(EvalContext *ctx,
...
+                                        bool actuate_ifelse)
 {
     PromiseResult result = PROMISE_RESULT_SKIPPED;
 
+    /* In the case of ifelse() we must always include an extra round of "actuation"
+     * in the while loop below. PromiseIteratorNext() will return false in the case
+     * that there are doubly-unresolved Rvals like $($(missing)).
+     * We can't add an empty wheel because that is skipped as well as noted in
+     * libpromises/iteration.c ShouldAddVariableAsIterationWheel(). */
+    bool ifelse_actuated = ! actuate_ifelse;
```
Comment (craigcomstock) [inline libpromises/expand.c:201]: @vpodzime especially this comment above here. This describes my understanding. Might be a bit long but figured I'd start with too much information. :)

**Comment by craigcomstock** [comment]
Yes. Without the C change I get the following failure with the ifelse.cf test :)

```
R: --- /northern.tech/cfengine/core/tests/acceptance/./01_vars/02_functions/ifelse.cf.expected.json     2020-04-23 16:14:52.200186967 +0000
+++ /northern.tech/cfengine/core/tests/acceptance/workdir/__01_vars_02_functions_ifelse_cf/tmp/TESTDIR.cfengine/actual  2020-04-23 16:29:31.078739773 +0000
@@ -6,6 +6,5 @@
   "hardclass": "hardclass OK",
   "one": "1",
   "single": "single string parameter",
-  "unresolved1": "expected fallback 1",
-  "unresolved2": "expected fallback 2"
+  "unresolved1": "expected fallback 1"
 }
R: /northern.tech/cfengine/core/tests/acceptance/./01_vars/02_functions/ifelse.cf FAIL
```

So I would say the test is good. :+1:

File: libpromises/expand.c (hunk 4 above)
Comment (vpodzime) [reply libpromises/expand.c:201]: Yes, now it makes much more sense. Thanks! Just a nitpick: `!actuate_ifelse`

**Comment by craigcomstock** [comment]
Ok @vpodzime: nitpick fixed and commits squashed!

File: libpromises/expand.c (hunk 1 above)
Comment (craigcomstock) [reply libpromises/expand.c]: fixed

**Comment by craigcomstock** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5053/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5053/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5053/

**Pull request merged**